### ISSUE TITLE
Server can't process add operations for tables with required fields when hitting page size limit

### ISSUE THREAD

**Issue by Toccata6767** [stefffdev/NubeSync#28]
When the server is processing operations, it first groups the operations by TableName and ItemId. It then determines what the operation type is (Added, Modified, Deleted), each resulting in a different behavior. When adding a new item there will always be one Added operation in the group along with several Modified operations. The server creates the new item and then modifies it according to the operations in the group. It then calls the AddAsync method from the DbContext with this new item. Even though these modification operations are separate from the add operations, since they are executed together, there shouldn't be any problems with required fields. The modification operations set the data for the required field before the server attempts to add it to the database through the DbContext.

There is a problem that can occur when the client pushes these operations to the server where the size of the batch is equal to the page size of 100. If the Add operation is near the end of the list it is is possible that the Modified operations for some required fields may be left out of the batch. The server processes the operations and tries to add the new item, an exception will be thrown because of the null value in the required field. The client gets back a BadRequest, and so none of the batch of operations that were completed get deleted. Next time the changes are pushed, the batch will be exactly the same and the problem will repeat. 

There are probably several ways this could be handled. What I ended up doing was rewriting the GetOperationsAsync method in NubeSQLiteDataStore.Operations.cs file. It will now only include Add operations for an item if it can also fit all of the Modification operations for that item as well.

```C#
public async Task<IQueryable<NubeOperation>> GetOperationsAsync(int numberOfOperations = 0)
{
    var groupedOperations = (await Database.Table<NubeOperation>().OrderBy(o => o.CreatedAt).ToListAsync()).GroupBy(o => new { o.TableName, o.ItemId });

    var count = 0;
    var operations = new List<NubeOperation>();

    foreach (var group in groupedOperations)
    {
        if (numberOfOperations != 0 && count >= numberOfOperations)
            break;

        count += group.Count();
        if (group.Any(g => g.Type == OperationType.Added) && numberOfOperations != 0 && count > numberOfOperations)
            break;

        operations.AddRange(group);
    }

    return operations.AsQueryable();
}
```

**Comment by stefffdev** [comment]
Thanks for the detailed report, I have not run into this one yet! 👍
Your fix is looking good to me, I will have a look at it soon.

Out of curiosity, may I ask what you are using nubesync for?

**Comment by Toccata6767** [comment]
I am writing a iOS app using Xamarin.Forms for a juvenile detention center. The app itself is fairly basic, they need to manage juveniles, staff, and any commissary related stuff. They don't always have an active internet connection and so they need a way for the app to still work with local data and then changes be synced when the internet connection returns. The offline/online sync is a very important feature for them so it is nice that you made NubeSync available like this. It saves me a ton of time from attempting to write my own. Thanks again!

**Comment by stefffdev** [comment]
Thanks for sharing your use case, always interesting to hear such stories :)
This issue should be fixed with the current release (1.5.2).
